File: web/src/lib/landmarks.ts
// Fetching the landmarks a short form can be anchored to.
//
// The shards are built by scripts/build-landmarks.mjs and keyed by level-3
// cell -- the first three characters of a landmark's own code. Which shards a
// lookup needs is decided by the recovery box of section 12.3, never by a
// radius: see reference.ts for why that distinction is not cosmetic.

import { GPC } from '@pranavpatel.ca/algo-gridpointcode';
import { RECOVERY, references, type Candidate, type Landmark } from './reference';

/** The order is the file format's, written as an index by the build script. */
const KINDS = ['structure', 'natural', 'place'] as const;
export type Kind = (typeof KINDS)[number];

export interface Choice extends Candidate {
    kind: Kind;
    region: string;
    /**
     * The landmark shares the point's level-5 cell.
     *
     * That is worth marking because it is a stronger promise than the box. A
     * listener who has this data can take the landmark's cell and use its
     * centre, and recovery is then exact by construction rather than dependent
     * on anyone's coordinates agreeing -- which is the weakness section 12.3 is
     * candid about and the one no arithmetic here can repair.
     */
    exact: boolean;
}

interface Shard {
    regions: string[];
    landmarks: [string, number, number, number, number][];
}

// A shard once fetched is kept: the data does not change between deployments,
// and a reader nudging a point around asks for the same three characters over
// and over. A miss is remembered too -- most of the planet is ocean, and a
// shard that does not exist should be asked for once, not on every keystroke.
const held = new Map<string, Promise<Shard | null>>();

const base = () => import.meta.env.BASE_URL.replace(/\/$/, '');
const url = (shard: string) => `${base()}/landmarks/${shard}.json`;

function fetchShard(shard: string): Promise<Shard | null> {
    const already = held.get(shard);
    if (already) return already;

    const pending = fetch(url(shard))
        .then((response) => (response.ok ? (response.json() as Promise<Shard>) : null))
        .catch(() => null);          // offline, or no shard there: the same answer

    held.set(shard, pending);
    return pending;
}

interface Manifest {
    level: number;
    shards: number;
    landmarks: number;
    /** When the archive was built. What tells a held copy from a current one. */
    built: string;
}

// Which level the shards were cut at is a property of the data, not of this
// file. Assuming it means that rebuilding at a different level turns every
// lookup into a miss, and a miss reads as `no landmarks near here` rather than
// as the broken deployment it is.
let described: Promise<Manifest | null> | null = null;

function manifest(): Promise<Manifest | null> {
    described ??= fetch(`${base()}/landmarks/manifest.json`)
        .then((response) => (response.ok ? (response.json() as Promise<Manifest>) : null))
        .catch(() => null);
    return described;
}

/**
 * The shards the recovery box reaches into.
 *
 * Taken from the corners rather than the centre, because a box that straddles
 * a boundary is exactly the case a centre lookup would miss -- and missing one
 * does not raise anything, it just quietly shortens the list.
 */
export function shardsFor(latitude: number, longitude: number, level: number): string[] {
    const shards = new Set<string>();
    for (const dy of [-1, 1]) {
        for (const dx of [-1, 1]) {
            const corner = latitude + dy * RECOVERY.latitude;
            const wrapped = ((longitude + dx * RECOVERY.longitude + 540) % 360) - 180;
            try {
                shards.add(GPC.cell(GPC.encode(Math.max(-90, Math.min(90, corner)), wrapped), level));
            } catch {
                // A corner past the pole or inside the reserved range has no
                // shard. The other corners still do, and the box is clipped to
                // what the world actually holds.
            }
        }
    }
    return [...shards];
}

/**
 * Every landmark this point's short form may be given as its reference,
 * nearest first.
 *
 * An empty result is an answer rather than a failure, and the caller is
 * expected to say so: in open country there is no landmark near enough, and
 * the honest response is the full ten characters, which is the form of record
 * in any case.
 */
export async function nearby(latitude: number, longitude: number): Promise<Choice[]> {
    const described = await manifest();

    // Not an empty list. An empty list means open country, and saying that
    // when the archive simply is not deployed would be a lie the reader
    // cannot see through -- they would believe there is nowhere near them.
    if (!described) throw new Error('no landmark data is deployed');

    const shards = await Promise.all(
        shardsFor(latitude, longitude, described.level).map(fetchShard),
    );

    let here: string | null = null;
    try {
        here = GPC.cell(GPC.encode(latitude, longitude), 5);
    } catch {
        return [];                   // a point with no code has no short form
    }

    const found: (Landmark & { kind: Kind; region: string })[] = [];
    for (const shard of shards) {
        if (!shard) continue;
        for (const [name, lat, lng, region, kind] of shard.landmarks) {
            found.push({
                name,
                latitude: lat,
                longitude: lng,
                region: shard.regions[region] ?? '',
                kind: KINDS[kind] ?? 'place',
            });
        }
    }

    return references(latitude, longitude, found).map((candidate) => {
        const source = candidate as Candidate & { kind: Kind; region: string };
        let exact = false;
        try {
            exact = GPC.cell(GPC.encode(candidate.latitude, candidate.longitude), 5) === here;
        } catch {
            exact = false;
        }
        return { ...source, exact };
    });
}

/** The reference line a reader can copy, said the way section 12.1 writes it. */
export function anchored(short: string, choice: Choice): string {
    const dashed = short.startsWith('-') ? short : `-${short}`;
    return `${dashed} near ${choice.name}, ${choice.region}`;
}

// ── keeping an area to hand ────────────────────────────────────────────────

/**
 * The alphabet of section 4, written out rather than read from the library,
 * which keeps it private. It cannot change without changing the format, so
 * this is a constant in the same sense the grid is.
 */
const ALPHABET = '0123456789CDFGHJKLMNPRTWX';

/**
 * Two stores, because two different things are being remembered.
 *
 * This one holds what a reader deliberately asked for, and is the only one this
 * file writes or counts. The worker keeps a second store of whatever it served
 * along the way -- a free convenience, and nobody's decision -- which is never
 * reported, because telling someone an area is held when they merely glanced at
 * it is the same sentence as telling them it is ready for a journey, and only
 * one of those is true.
 *
 * Named for the build, so a new archive does not inherit an old one's copies.
 */
const keptName = (described: Manifest) => `gpc-kept-${described.built}`;

/** Whether the browser will let anything be kept at all. */
export const canKeep = () => typeof caches !== 'undefined';

export interface Area {
    /** The cell being kept: one level up from a shard. */
    cell: string;
    /** How many of its shards hold anything. Most of the planet is ocean. */
    shards: number;
    bytes: number;
}

/**
 * Fetch and keep every shard of the area around a point.
 *
 * The area is the cell one level above a shard -- about 200 by 267 km at the
 * level the archive is cut for. That is a deliberate size: large enough to be
 * worth asking for before a journey, small enough that the answer is a few
 * hundred kilobytes rather than the eighty-odd megabytes the whole world
 * weighs. Shards already held are not fetched again.
 *
 * Written into the cache the worker reads from, under the name the current
 * build stamped, so keeping something and being served it offline are the
 * same act rather than two that have to agree.
 */
export async function keepArea(latitude: number, longitude: number): Promise<Area> {
    const described = await manifest();
    if (!described) throw new Error('no landmark data is deployed');
    if (!canKeep()) throw new Error('this browser will not keep anything');

    const area = GPC.cell(GPC.encode(latitude, longitude), described.level - 1);
    const cache = await caches.open(keptName(described));
    const wanted = [...ALPHABET].map((symbol) => url(area + symbol));

    // Fetching happens together, because it is all network waiting and there
    // are at most twenty-five of them.
    await Promise.all(
        wanted.map(async (address) => {
            if (await cache.match(address)) return;

            let response: Response;
            try {
                response = await fetch(address);
            } catch {
                return;                  // no network for this one; the rest may work
            }
            // A 404 is ocean, which is an answer rather than a failure.
            if (!response.ok) return;

            await cache.put(address, response);
        }),
    );

    // Measuring happens one at a time, and that is not fussiness. Reading
    // twenty-five cached bodies at once returned 1,177 bytes where reading the
    // same twenty-five in turn returned 83,506 -- the stored data was whole
    // either way, but concurrent reads of it were not. A wrong size is a small
    // lie told confidently, which is worse than the millisecond this costs.
    let shards = 0;
    let bytes = 0;
    for (const address of wanted) {
        const held = await cache.match(address);
        if (!held) continue;
        shards += 1;
        bytes += (await held.arrayBuffer()).byteLength;
    }

    return { cell: area, shards, bytes };
}

/**
 * How much of the area around one point is already held.
 *
 * Separate from `kept` below, which counts everything anywhere. The difference
 * matters to a reader: being told that twenty-five shards are held is no use
 * when standing somewhere none of them cover, and is worse than no message at
 * all, because it reads as reassurance.
 */
export async function heldArea(
    latitude: number,
    longitude: number,
): Promise<{ cell: string; shards: number } | null> {
    if (!canKeep()) return null;

    const described = await manifest();
    if (!described) return null;

    const area = GPC.cell(GPC.encode(latitude, longitude), described.level - 1);
    const name = keptName(described);
    if (!(await caches.keys()).includes(name)) return { cell: area, shards: 0 };

    const cache = await caches.open(name);
    let shards = 0;
    for (const symbol of ALPHABET) {
        if (await cache.match(url(area + symbol))) shards += 1;
    }
    return { cell: area, shards };
}

/** How much is being kept, across every area asked for so far. */
export async function kept(): Promise<{ shards: number } | null> {
    if (!canKeep()) return null;
    const described = await manifest();
    if (!described) return null;

    const names = await caches.keys();
    const name = keptName(described);
    if (!names.includes(name)) return { shards: 0 };

    const cache = await caches.open(name);
    return { shards: (await cache.keys()).length };
}

/**
 * Give it all back: what was kept and what was merely passed through, from
 * every build rather than only the current one.
 *
 * Both, because a reader pressing this wants the space back and does not care
 * which of our two stores happens to hold it. The runtime store fills again on
 * its own as they look around, which is invisible and costs nothing -- and,
 * because only the kept store is ever counted, it will not be mistaken for
 * something they asked for.
 */
export async function forget(): Promise<void> {
    if (!canKeep()) return;
    for (const name of await caches.keys()) {
        if (name.startsWith('gpc-kept-') || name.startsWith('gpc-landmarks-')) {
            await caches.delete(name);
        }
    }
}

File: android/core/src/main/kotlin/com/gridpointcode/core/Landmarks.kt
package com.gridpointcode.core

import ca.pranavpatel.algo.gridpointcode.GPC
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.sqrt

/*
 * Which landmarks a short form can be given as its reference.
 *
 * A short form is five characters and a place near the point: the listener
 * finds the place, and recovery returns the full code. Section 12.3 bounds when
 * that works, and these rules are the website's own (web/src/lib/reference.ts
 * and landmarks.ts), read from the same archive of landmarks.
 */

/**
 * The recovery bound of section 12.3: within half a level-5 cell in each axis,
 * 0.03598848 degrees of latitude and 0.04798464 of longitude.
 *
 * A box in degrees, and the distinction is not pedantry. The longitude bound is
 * a fixed number of degrees, so the ground it covers shrinks with the cosine of
 * the latitude: a landmark three kilometres due east is well inside the bound
 * at the equator and outside it from about 60 degrees up, which is Oslo,
 * Helsinki and most of Alaska. Anything filtering by a radius in metres is
 * wrong there, and wrong silently: recovery returns another cell's copy of the
 * same five characters, a plausible place eight or ten kilometres away.
 */
object Recovery {
    // Counted in finest rows and columns, as the specification states it: a
    // level-5 cell is 3,125 of them, so half of it is 1,562 whole ones. Exactly
    // half a level-5 cell in degrees is half a finest cell too generous, and
    // that is the half a landmark on the edge of the box would fall into.
    private val finest = GPC.CellDimensions(10)
    private val rows = (GPC.CellDimensions(5).LatitudeSpan / finest.LatitudeSpan).roundToInt() / 2

    val latitude: Double = rows * finest.LatitudeSpan
    val longitude: Double = rows * finest.LongitudeSpan
}

/** What kind of thing a landmark is, in the archive's own order. */
enum class LandmarkKind { STRUCTURE, NATURAL, PLACE }

/** A named place from the archive: named uniquely within its region, so the pair is unambiguous. */
data class Landmark(
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val region: String,
    val kind: LandmarkKind,
)

/**
 * A landmark a short form may be given with, as the reader is shown it.
 *
 * @property metres great-circle distance, for the reader and never for the rule
 * @property bearing which way it lies, to the nearest of eight points
 * @property tightness how much of the recovery box it uses: 0 at the point, 1 on the edge
 * @property exact it shares the point's level-5 cell, so a listener who resolves
 *   it by that cell's centre recovers the code whatever anyone's coordinates say
 */
data class Anchor(
    val landmark: Landmark,
    val metres: Double,
    val bearing: String,
    val tightness: Double,
    val exact: Boolean,
)

/**
 * How much of the recovery box a reference uses up. Whichever axis is nearer
 * its limit decides, because either one failing is enough.
 */
fun tightness(point: Point, latitude: Double, longitude: Double): Double = max(
    abs(latitude - point.latitude) / Recovery.latitude,
    abs(wrap(longitude - point.longitude)) / Recovery.longitude,
)

/**
 * The landmarks a short form may be paired with, nearest first.
 *
 * Filtered by the box and sorted by distance, two different measures on
 * purpose: the box is what the specification guarantees, and distance is what a
 * reader understands. Sorting by the box's own fraction would put a landmark
 * further away above a nearer one whenever the axes disagree, which is correct
 * and reads as a mistake.
 */
fun anchorsFor(point: Point, landmarks: Iterable<Landmark>): List<Anchor> {
    val here = runCatching { GPC.Cell(GPC.Encode(point.latitude, point.longitude, false), 5) }
        .getOrNull() ?: return emptyList()
    return landmarks
        .mapNotNull { landmark ->
            val tight = tightness(point, landmark.latitude, landmark.longitude)
            if (tight > 1) return@mapNotNull null
            val theirs = runCatching { GPC.Cell(GPC.Encode(landmark.latitude, landmark.longitude, false), 5) }
                .getOrNull()
            Anchor(
                landmark = landmark,
                metres = metres(point, landmark.latitude, landmark.longitude),
                bearing = bearing(point, landmark.latitude, landmark.longitude),
                tightness = tight,
                exact = theirs == here,
            )
        }
        .sortedBy { it.metres }
}

/**
 * The archive's shards the recovery box reaches into, named by the cell at the
 * archive's level.
 *
 * Taken from the box's four corners rather than its centre, because a box that
 * straddles a boundary is exactly what a centre lookup misses, and missing a
 * shard raises nothing: it just quietly shortens the list.
 */
fun shardsFor(point: Point, level: Int): List<String> {
    val shards = linkedSetOf<String>()
    for (dy in intArrayOf(-1, 1)) {
        for (dx in intArrayOf(-1, 1)) {
            val latitude = (point.latitude + dy * Recovery.latitude).coerceIn(-90.0, 90.0)
            val longitude = wrap(point.longitude + dx * Recovery.longitude)
            // A corner past the pole or in the reserved range has no shard. The
            // others still do, and the box is clipped to what the world holds.
            runCatching { GPC.Cell(GPC.Encode(latitude, longitude, false), level) }
                .onSuccess { shards += it }
        }
    }
    return shards.toList()
}

/**
 * The line a reader shares, as section 12.1 writes it: the short form with its
 * leading dash and no hash, then the place. Code first, because a dash at the
 * end reads as a minus sign; "near", because the point can be kilometres off;
 * the region spelled out, because it is what tells two places of one name apart.
 */
fun anchored(code: String, landmark: Landmark): String {
    val place = if (landmark.region.isEmpty()) landmark.name else "${landmark.name}, ${landmark.region}"
    return "-${GPC.Shorten(code)} near $place"
}

/** A longitude difference carried into the range holding the short way round. */
private fun wrap(degrees: Double): Double = ((degrees + 540) % 360) - 180

private const val EARTH_METRES = 6371008.8

private fun radians(degrees: Double) = degrees * PI / 180

private fun metres(point: Point, latitude: Double, longitude: Double): Double {
    val dLat = radians(latitude - point.latitude)
    val dLng = radians(wrap(longitude - point.longitude))
    val a = sin(dLat / 2).let { it * it } +
        cos(radians(point.latitude)) * cos(radians(latitude)) * sin(dLng / 2).let { it * it }
    return 2 * EARTH_METRES * asin(min(1.0, sqrt(a)))
}

/**
 * Not the nudge pad's [Compass]. The two hold the same eight names and mean
 * different things: that one is the order the library returns neighbouring
 * cells in, this one is a direction on the ground.
 */
private val OCTANTS = listOf("N", "NE", "E", "SE", "S", "SW", "W", "NW")

private fun bearing(point: Point, latitude: Double, longitude: Double): String {
    val dLng = radians(wrap(longitude - point.longitude))
    val y = sin(dLng) * cos(radians(latitude))
    val x = cos(radians(point.latitude)) * sin(radians(latitude)) -
        sin(radians(point.latitude)) * cos(radians(latitude)) * cos(dLng)
    val degrees = atan2(y, x) * 180 / PI
    return OCTANTS[(((degrees + 360) % 360) / 45).roundToInt() % 8]
}

/** How many shards the kept area reaches out from the place's own, every way. */
private const val AREA_REACH = 2

/**
 * The shards of the area kept around a point: its own shard and two more every
 * way, a block of twenty-five, 200 km north to south.
 *
 * Centred on the place rather than a cell of the grid. A cell one level up is
 * the same size, and is what the website keeps, but it is fixed to the grid:
 * downtown Toronto sits in the corner of its cell, which reaches 190 km west and
 * stops 15 km east. Large enough to be worth keeping before a journey, small
 * enough to be a few hundred kilobytes rather than the eighty-odd megabytes the
 * whole archive weighs.
 *
 * Found by stepping whole shards from the point, which lands in each neighbour
 * at the point's own place within it. Fewer than twenty-five near a pole, where
 * the rows run out.
 */
fun areaAround(point: Point, level: Int): List<String> {
    val cell = GPC.CellDimensions(level)
    val shards = linkedSetOf<String>()
    for (row in -AREA_REACH..AREA_REACH) {
        for (column in -AREA_REACH..AREA_REACH) {
            val latitude = point.latitude + row * cell.LatitudeSpan
            if (latitude < -90 || latitude > 90) continue
            val longitude = wrap(point.longitude + column * cell.LongitudeSpan)
            runCatching { GPC.Cell(GPC.Encode(latitude, longitude, false), level) }
                .onSuccess { shards += it }
        }
    }
    return shards.toList()
}

/** The kept area's size on the ground at a latitude, in metres: north to south, then east to west. */
fun areaMetres(latitude: Double, level: Int): Pair<Double, Double> {
    val cell = GPC.CellDimensions(level)
    val across = 2 * AREA_REACH + 1
    return across * cell.NorthSouth to across * cell.EastWest * cos(radians(latitude))
}
